db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/BinaryComparisonOperatorNode.java

```diff
@@ -55,154 +55,101 @@
-
-		/*
-		 * If we are comparing a non-string with a string type, then we
-		 * must prevent the non-string value from being used to probe into
-		 * an index on a string column. This is because the string types
-		 * are all of low precedence, so the comparison rules of the non-string
-		 * value are used, so it may not find values in a string index because
-		 * it will be in the wrong order. So, cast the string value to its
-		 * own type. This is easier than casting it to the non-string type,
-		 * because we would have to figure out the right length to cast it to.
-		 */
-		if (! leftTypeId.isStringTypeId() && rightTypeId.isStringTypeId())
-		{
-			if (leftTypeId.isBooleanTypeId() || leftTypeId.isDateTimeTimeStampTypeId()) {
-				rightOperand = (ValueNode)
-						getNodeFactory().getNode(
-								C_NodeTypes.CAST_NODE,
-								rightOperand,
-								new DataTypeDescriptor(
-										leftTypeId,
-										true,leftOperand.getTypeServices().getMaximumWidth()),
-								getContextManager());
-				((CastNode) rightOperand).bindCastNodeOnly();
-				rightTypeId = rightOperand.getTypeId();
-			}
-			else if (leftTypeId.isNumericTypeId() &&
-				     rightTypeId.isCharOrVarChar()) {
-
-				rightOperand = (ValueNode) getNodeFactory().getNode(
-					C_NodeTypes.CAST_NODE,
-					rightOperand,
-					DataTypeDescriptor.getBuiltInDataTypeDescriptor(Types.DOUBLE),
-					getContextManager());
-				((CastNode) rightOperand).bindCastNodeOnly();
-				rightTypeId = rightOperand.getTypeId();
-			}
-		}
-		else if (! rightTypeId.isStringTypeId() && leftTypeId.isStringTypeId())
-		{
-			if (rightTypeId.isBooleanTypeId() || rightTypeId.isDateTimeTimeStampTypeId()) {
-				leftOperand =  (ValueNode)
-						getNodeFactory().getNode(
-								C_NodeTypes.CAST_NODE,
-								leftOperand,
-								new DataTypeDescriptor(
-										rightTypeId,
-										true,
-										rightOperand.getTypeServices().getMaximumWidth()),
-								getContextManager());
-				((CastNode) leftOperand).bindCastNodeOnly();
-				leftTypeId = leftOperand.getTypeId();
-			}
-			else if (rightTypeId.isNumericTypeId() &&
-				     leftTypeId.isCharOrVarChar()) {
-
-				leftOperand = (ValueNode) getNodeFactory().getNode(
-					C_NodeTypes.CAST_NODE,
-					leftOperand,
-					DataTypeDescriptor.getBuiltInDataTypeDescriptor(Types.DOUBLE),
-					getContextManager());
-				((CastNode) leftOperand).bindCastNodeOnly();
-				leftTypeId = leftOperand.getTypeId();
-			}
-		}
+        TypeId rightTypeId = rightOperand.getTypeId();
+
+        if (! leftTypeId.isStringTypeId() && rightTypeId.isStringTypeId())
+        {
+            rightOperand = addCastNodeForStringToNonStringComparison(leftOperand, rightOperand);
+            rightTypeId = rightOperand.getTypeId();
+        }
+        else if (! rightTypeId.isStringTypeId() && leftTypeId.isStringTypeId())
+        {
+            leftOperand.addCastNodeForStringToNonStringComparison(rightOperand, leftOperand);
```

Comment:
Should this be like this instead?
```suggestion
            leftOperand = addCastNodeForStringToNonStringComparison(rightOperand, leftOperand);
``` 
Is it possible we are not testing this branch? Can we add a test?

Reply:
Good point, I will add a test and fix that.